File: percolation.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""
import array
class WeightedQuickUnion:
    def __init__(self, N):
        self.count = N
        self.id = array.array('i', range(N))
        self.size = array.array('i', [1] * N)
    
    def connected(self, p, q):
        return self.find(p) == self.find(q)
    
    def count(self):
        return self.count
    
    def find(self, p):
        while (p != self.id[p]):
            p = self.id[p]
        return p
    
    def get(self, p):
        return self.id[p]
    
    def union(self, p, q):
        rootP = self.find(p)
        rootQ = self.find(q)
        if (rootP == rootQ):
            return
        
        if (self.size[rootP] < self.size[rootQ]):
            self.id[rootP] = rootQ
            self.size[rootQ] += self.size[rootP]
        else:
            self.id[rootQ] = rootP
            self.size[rootP] += self.size[rootQ]
        self.count -= 1


# %% Percolation
class Percolation:
    def __init__(self, N):
        self.N = N
        self.uf = WeightedQuickUnion(N*N+2)
        self.state = array.array('b', [False] * (N*N+2))
        self.vTop = 0
        self.vBottom = N*N+1
        
        last_row = N * (N-1)
        for k in range(1, N+1):
            self.uf.union(last_row + k, self.vBottom)
            self.uf.union(k, self.vTop)
    
    def open(self, p):
        self.state[p] = True
        
        self.unionNeighbor(p, p - 1)
        self.unionNeighbor(p, p - self.N)
        self.unionNeighbor(p, p + self.N)
        self.unionNeighbor(p, p + 1)
    
    def isOpen(self, p):
        return self.state[p]
    
    def isFull(self, p):
        return self.isOpen(p) and self.uf.connected(self.vTop, p)
    
    def percolates(self):
        return self.uf.connected(self.vTop, self.vBottom)
    
    def valid(self, p):
        if (p <= self.vTop or p >= self.vBottom):
            return False
        
        i,j = self.coordinates(p)        
        if (i <= 0 or j <= 0):
            return False
        else:
            return (i <= self.N and j <= self.N)
    
#    def xyTo1D(self, i, j):
#        N = len(self.grid)
#        # Error checking with valid
#        return (i-1) * N + j
    
    def coordinates(self, p):
        return (p // self.N, p % self.N)
    
    def unionNeighbor(self, p, q):
        if (self.valid(q) and self.isOpen(q)):
            self.uf.union(p, q)


# %% Percolation Stats
import math
import random

class PercolationStats:
    def __init__(self, N, T):
        self.N = N
        self.trials = T
        self.run = [self.simulation(N) for x in range(T)]
        self.mu = sum(self.run) / T
        self.sd = sum([(r-self.mu)**2 for r in self.run]) / T
    
    def mean(self):
        return self.mu
    
    def stddev(self):
        return self.sd
    
    def confidenceLo(self):
        return self.mu - ((1.96 * self.sd) / math.sqrt(self.trials))
    
    def confidenceHi(self):
        return self.mu + ((1.96 * self.sd) / math.sqrt(self.trials))

    def simulation(self, N):
        Perc = Percolation(N)
        count = 0
        while (not Perc.percolates()):
            p = random.randint(1, N*N)
            if (not Perc.isOpen(p)):
                Perc.open(p)
                count += 1
        
        return count * 1.0 / (N**2)

    def __repr__(self):
        return "Percolation Simulation of size {} for {} trials".format(self.N, self.trials)

    def __str__(self):
        return "({:.6}, {:.6}, {:.6})".format(self.confidenceLo(), self.mean(), self.confidenceHi())

File: Percolation.java
/*****************************************************************************
 * Author:       Bryan Goodrich
 * Written:      2/4/2015
 * Last updated: 2/8/2015
 * 
 * Compilation:  javac Percolation.java
 * Execution:    java Percolation input.txt
 * Dependencies: StdOut.java In.java
 * 
 * Models a percolating system to be used in simulations or exploration.
 * 
 ****************************************************************************/

/**
 * <i>Percolation</i> Model.
 * 
 * This class provides methods for modeling a grid based percolating system.
 * The API here is designed by Robert Sedgewick and Kevin Wayne as part of
 * their online <i>Coursera</i> class.
 * 
 * @author Bryan Goodrich
 */
public class Percolation {
    
    // positional pointers to virtual top and bottom of UF data structure
    private final int vTop, vBottom;
    
    // a union-find data structure for maintaining connections in system
    private WeightedQuickUnionUF uf;
    
    // a 2D array for managing system state (open/closed)
    private boolean[][] grid;
    
    /**
     * Creates a percolation system
     * 
     * The model is a square NxN state grid that uses a union-find
     * data structure to manage the site (cell) connections. The top
     * and bottom layers of the grid are attached to virtual sites in
     * the union-find structure at initialization, providing for easier
     * ways to identify percolation across the system
     * 
     * @param N  the size of the square state grid for this model
     */    
    public Percolation(int N) {
        if (N <= 0) 
            throw new IllegalArgumentException("Invalid argument: non-positive N");
        this.uf = new WeightedQuickUnionUF(N*N+2);
        this.grid = new boolean[N][N];
        this.vBottom = N*N+1;
        this.vTop = 0;
        
        // Connect the bottom and top rows to their virtual links
        if (N > 1) {
            for (int k = 1; k <= N; k++) {
                this.uf.union(xyTo1D(N, k), vBottom);
                this.uf.union(xyTo1D(1, k), vTop);
            }
        }
    }
    
    /**
     * Opens a site at the given coordinates
     * 
     * Opening a site includes both changing its grid state as well as
     * connecting it in the union-find structure to its grid neighbors above,
     * below, left, and right if those neighbors are also open.
     * 
     * @param i  x-coordinate of site beginning from northwest corner
     * @param j  y-coordinate of site beginning from northwest corner
     */    
    public void open(int i, int j) {
        grid[i-1][j-1] = true;
        
        if (grid.length == 1) {
            this.uf.union(xyTo1D(i, j), vBottom);
            this.uf.union(xyTo1D(i, j), vTop);
        }
        
        int p = xyTo1D(i, j);
        unionNeighbor(p, i-1, j);  // above
        unionNeighbor(p, i, j-1);  // left
        unionNeighbor(p, i, j+1);  // right
        unionNeighbor(p, i+1, j);  // below
    }
    
    /**
     * Checks if a given site is open
     * 
     * @param i  x-coordinate of site beginning from northwest corner
     * @param j  y-coordinate of site beginning from northwest corner
     * @return   true if grid states this site is open
     */    
    public boolean isOpen(int i, int j) { return grid[i-1][j-1]; }
    
    /**
     * Checks if a given site is full (percolates to these coordinates)
     * 
     * @param i  x-coordinate of site beginning from northwest corner
     * @param j  y-coordinate of site beginning from northwest corner
     * @return   true if the site is open and connected to the top
     */
    public boolean isFull(int i, int j) {
        return isOpen(i, j) && uf.connected(vTop, xyTo1D(i, j));
    }
    
    /**
     * Checks if the system percolates from top to bottom
     * 
     * @return  true if the virtual top and bottom are connected as
     *          represented in the union-find data structure
     */
    public boolean percolates() { return uf.connected(vTop, vBottom); }
    
    /**
     * Test client for executing a model from an input file
     * 
     * The file must begin with grid size N on the first line.
     * Proceeding lines must include pairs of integers separated by
     * whitespaces for which cells in the grid to open
     */
    public static void main(String[] args) {
        In in = new In(args[0]);      // input file
        int N = in.readInt();         // N-by-N percolation system
        
        Percolation perc = new Percolation(N);
        
        while (!in.isEmpty()) {
            int i = in.readInt();
            int j = in.readInt();
            perc.open(i, j);
            StdOut.println(i + " " + j);
        }
    }
    
    /**
     * Validates the index is in range
     * 
     * @param i  index to be validated
     * @return true if index between 1 and N, the size of one side of the system
     */
    private boolean valid(int i) {
        int N = grid.length;
        return (i > 0 && i <= N); 
    }
    
    /**
     * Converts grid coordinates to union-find 1-dimensional representative
     * 
     * @param i the x-coordinate beginning from the northwest corner
     * @param j the y-coordinate beginning from the northwest corner
     * @throws  a IndexOutOfBoundsException if index not bound to grid system [1,N]
     * @return  an integer cooresponding to the union-find array position of
     *          this coordinate
     */
    private int xyTo1D(int i, int j) {
        int N = grid.length;
        if (!valid(i)) 
            throw new IndexOutOfBoundsException("row index i out of bounds");
        if (!valid(j)) 
            throw new IndexOutOfBoundsException("column index j out of bounds");
        return (i-1) * N + j;
    }
    
    /**
     * Unions a given point to its open neighbors at the given coordinates
     * 
     * @param p  the union-find position that was opened
     * @param i  the x-coordinate of a neighbor to be joined
     * @param j  the y-coordinate of a neighbor to be joined
     * @see      open
     * @see      xyTo1D
     */ 
    private void unionNeighbor(int p, int i, int j) {
        if (!valid(i) || !valid(j)) return;
        
        if (isOpen(i, j))
            uf.union(p, xyTo1D(i, j));
    }
}
